content/blog/2025-06-30-Zenoh-Pico-Peer-to-peer-unicast.md

```diff
@@ -0,0 +1,136 @@
+
+Architecture-wise, we use non-blocking sockets and I/O multiplexing to handle all connections on a single RX thread, plus an additional thread that listens on a socket and accepts incoming connections. For resource-efficiency reasons, peer-unicast nodes do not route traffic: every message received from a connected peer triggers our API, and every message created via our API is sent to all connected peers. This design allows for a single TX and a single RX buffer.
+
+## Examples:
+
+Here is an example showing how to implement a 1:N (or N:1) communication graph:
+
+{{< figure-inline
+    src="../../img/20250630-Zenoh-Pico-peer-to-peer-unicast/1-n.png"
+    class="figure-inline"
+    alt="1:N diagram" >}}
+
+If we assume a single publisher connected to 3 subscribers, here’s how we could configure it:
+
+ ```Bash
+ ./build/example/z_pub -l tcp/127.0.0.1:7447
+ ./build/example/z_sub -e tcp/127.0.0.1:7447
+ ./build/example/z_sub -e tcp/127.0.0.1:7447
+ ./build/example/z_sub -e tcp/127.0.0.1:7447
+ ```
+
+To implement an N:N graph:
+
+{{< figure-inline
+    src="../../img/20250630-Zenoh-Pico-peer-to-peer-unicast/n-n.png"
+    class="figure-inline"
+    alt="N:N diagram" >}}
+
+```Bash
+./build/example/z_pub -l tcp/127.0.0.1:7447
+./build/example/z_sub -l tcp/127.0.0.1:7448 -e tcp/127.0.0.1:7447
+./build/example/z_sub -l tcp/127.0.0.1:7449 -e tcp/127.0.0.1:7447 -e tcp/127.0.0.1:7448
+./build/example/z_sub -e tcp/127.0.0.1:7447 -e tcp/127.0.0.1:7448 -e tcp/127.0.0.1:7449
+```
+
+# Performances
+
+## Test Details
+
+In addition to enabling peer-to-peer unicast, we improved general library CPU utilization, further boosting throughput and latency by approximately 10%. The tests were run on an Ubuntu 22.04 laptop equipped with an AMD Ryzen 7735U and 32 GB of RAM.
+
+## Configuration
+
+Note that the Zenoh-Pico configuration used for testing deviates from the default. Here are the changes:
+
+* `Z_FEATURE_SESSION_CHECK` set to 0 (default 1): Skips the publisher’s session reference upgrade. This is risky if you use the publisher after closing the session.
+* `Z_FEATURE_BATCH_TX_MUTEX` set to 1 (default 0): Allows the batching mechanism to hold the mutex, which can prevent the lease task from sending keep-alives, triggering connection closure.
+* `Z_FEATURE_RX_CACHE` set to 1 (default 0): Activates the RX LRU cache. It consumes some memory to store results of key expressions that trigger callbacks—useful in repetitive, high-throughput scenarios.
+
+## Results
+
+{{< figure-inline
+    src="../../img/20250630-Zenoh-Pico-peer-to-peer-unicast/perf_lat.png"
+    class="figure-inline"
+    alt="P2p latency" >}}
+
+The round-trip latency for packets below 16 KiB is under 20 µs—meaning a one-way latency of under 10 µs. Peer-to-peer unicast delivers up to **70% lower latency** compared to client mode.
```

Comment:
Is this Round-trip time or latency? P2P latency and Round trip is not the same. Latency = RTT/2 can you please clarify?